src/common/cockpitchannel.c

```diff
@@ -158,6 +185,55 @@ on_transport_recv (CockpitTransport *transport,
   return TRUE;
 }
 
+static gboolean
+process_ping (CockpitChannel *self,
+              JsonObject *ping)
+{
+  GBytes *payload;
+
+  if (self->priv->throttled)
+    {
+      g_debug ("%s: received ping while throttled", self->priv->id);
+      g_queue_push_tail (self->priv->throttled, json_object_ref (ping));
+      return FALSE;
+    }
+  else
+    {
+      g_debug ("%s: replying to ping with pong", self->priv->id);
+      json_object_set_string_member (ping, "command", "pong");
+      payload = cockpit_json_write_bytes (ping);
+      cockpit_transport_send (self->priv->transport, NULL, payload);
+      g_bytes_unref (payload);
+      return TRUE;
+    }
+}
+
+static gboolean
+process_pong (CockpitChannel *self,
+              JsonObject *pong)
+{
+  gint64 sequence;
+
+  if (!cockpit_json_get_int (pong, "sequence", -1, &sequence))
+    {
+      g_message ("%s: received invalid \"pong\" \"sequence\" field", self->priv->id);
+      sequence = -1;
+    }
+
+  g_debug ("%s: received pong with sequence: %" G_GINT64_FORMAT, self->priv->id, sequence);
+  if (sequence > self->priv->out_window)
```

Comment:
does it make sense to check this value for plausibility, warning if we see something weird?